juju/juju
# refactor: move uniter facade to model config/info services

This PR moves the uniter facade over to using model config and info services instead of state.

A new `service.go` file has been added, containing the new interfaces `ModelConfigService` and `ModelInfoService`, and all existing services have been moved there as well.

Tests have been changed accordingly to use the service factory instead of state where appropriate. The LXD profile tests in particular can use mock services instead.

## Checklist




## Facade usage

The following uniter facade methods use model config, either directly or indirectly:
| method | config keys used |
|-|-|
| EnterScope | egress-subnets |
| NetworkInfo | egress-subnets |
| UpdateNetworkInfo | egress-subnets |
| CommitHookChanges | egress-subnets |

The following uniter facade methods use model info, either directly or indirectly:
| method | information used |
|-|-|
| ProviderType | cloud type |
| CanApplyLXDProfile | model type, cloud type |
| CurrentModel | name, uuid, type |
| CloudSpec | uuid |

## QA steps



<details>
<summary>Check that changes to `egress-subnets` are picked up by existing relations</summary>

Set up a CMR:
```
juju add-model src
juju deploy juju-qa-dummy-source
juju offer dummy-source:sink dummy-offer

juju add-model snk
juju deploy juju-qa-dummy-sink
juju relate dummy-sink src.dummy-offer
```
Wait for the apps to start up.

Check the egress subnets on the relation:
```
$ juju exec --unit dummy-sink/0 -- network-get source --egress-subnets
10.94.69.4/32
```

Set new `egress-subnets` value in model config, check that the egress subnets are changed in the relation.
```
$ juju model-config egress-subnets=1.1.1.1/32
$ juju exec --unit dummy-sink/0 -- network-get source --egress-subnets
1.1.1.1/32
```
</details>

<details>
<summary>Check that the `egress-subnets` value is picked up by new relations</summary>

Continuing from the state of the previous section:
```
$ juju deploy juju-qa-dummy-sink sink2
$ juju relate sink2 src.dummy-offer
$ juju exec --unit sink2/0 -- network-get source --egress-subnets
1.1.1.1/32
```
</details>

## Links



**Jira card:** JUJU-6101
